ext/pdo/pdo_dbh.c

```diff
@@ -1382,8 +1376,18 @@ static void dbh_free(pdo_dbh_t *dbh, bool free_persistent)
 		pefree((char *)dbh->persistent_id, dbh->is_persistent);
 	}
 
-	if (!Z_ISUNDEF(dbh->def_stmt_ctor_args)) {
-		zval_ptr_dtor(&dbh->def_stmt_ctor_args);
+	/* When the GC is running, the ctor_args will be added to the GC buffer marked to be released.
+	 * However, before freeing the buffer, it will release the object by calling zend_objects_store_del()
+	 * which calls pdo_dbh_free_storage() which in turn calls this function.
+	 * Thus we must not free the ctor_args when the GC is running. */
+	if (dbh->def_stmt_ctor_args) {
+		zend_gc_status gc_status;
+		zend_gc_get_status(&gc_status);
+
+		if (EXPECTED(!gc_status.active)) {
+			zend_array_release(dbh->def_stmt_ctor_args);
+			dbh->def_stmt_ctor_args = NULL;
+		}
```

Comment:
The GC avoids this issue by setting `GC_TYPE_INFO()` to `IS_NULL` [here](https://github.com/php/php-src/blob/7c4db150cb490b476c0b6a6bf9319306856b96ec/Zend/zend_gc.c#L1942) and [here](https://github.com/php/php-src/blob/7c4db150cb490b476c0b6a6bf9319306856b96ec/Zend/zend_hash.c#L1816) so that `zval_ptr_dtor()` is a no-op on that value when it's already being destroyed.

Checking `GC_TYPE(dbh->def_stmt_ctor_args)` before calling `zend_array_release(dbh->def_stmt_ctor_args)` would fix the issue:

```
	if (dbh->def_stmt_ctor_args && GC_TYPE(dbh->def_stmt_ctor_args) == IS_ARRAY) {
		zend_array_release(dbh->def_stmt_ctor_args);
```

Alternatively we could use `GC_DTOR()`, which is equivalent to `zval_ptr_dtor` for refcounted values, but it may unnecessarily call `gc_check_possible_root`. We could introduce a new macro `GC_DTOR_NOGC` that does not call ``gc_check_possible_root`, like `zval_ptr_dtor_nogc`.